crates/nu-command/src/conversions/into/duration.rs

```diff
@@ -409,7 +409,13 @@ fn convert_str_from_unit_to_unit(
 }
 
 fn string_to_duration(s: &str, span: Span, value_span: Span) -> Result<i64, ShellError> {
-    if let Some(expression) = parse_duration_bytes(s.as_bytes(), span) {
+    if let Some(Ok(expression)) = parse_unit_value(
```

Comment:
It looks a bit weird to have `parse_unit_value()`  require 2 level matching everywhere it is invoked.
Couldn't it have same contract as the top level `parse_*` do?  That is, return `Expression` and internally call `working_set.error()` to construct an error if the input doesn't parse as the expected shape?  Caller already tells you what the expected units are, you could add an argument for caller to pass in "filesize" or "duration" to plug into the `::LabelledError`.  Right now, it returns `Err(mk_error_for())` to caller, but the only thing captured in the closure is the string "filesize" or "duration", seems the long way around.

Reply:
1. the reason for the nesting is to handle [two sources of failure](https://github.com/nushell/nushell/blob/bf86cd50a5928d86dce0e9b5b52a269cec35595f/crates/nu-parser/src/parser.rs#L2337).
- errors in the number part of the value
- the value not having a valid unit, meaning it's not a `value-with-unit`

2. `parse_unit_value()` \*\*cannot\*\* take external state because it may be (and is) used in [contexts](https://github.com/nushell/nushell/blob/607136470d7c92f4867481c905c5b0fee3d6a0d8/crates/nu-command/src/conversions/into/duration.rs#L412) where we don't have access to `StateWorkingSet`

3. i read a comment on discord about future plans to support user defined units. those and even other inbuilt units just have to call this function, and `mk_err_for()` will make the error message unique to them. my argument for using `mk_err_for()` is that the caller has more context. i thought about adding another argument, but i decided to move some work to the caller. either is okay, i think. (though i don't really like functions with [too many arguments](https://rust-lang.github.io/rust-clippy/master/index.html#too_many_arguments), and `parse_unit_value()` is already not looking very good)